### PR TITLE
Federation: Improve the handling of service deletion in underlying clusters

### LINKED ISSUE #27623: Add e2e test that federated servicecontroller recreates deleted services in underlying clusters
As mentioned in https://github.com/kubernetes/kubernetes/pull/27587#discussion_r67451976 when the service is deleted from the underlying cluster, the controller should recreate it, but it does not.

Below's a transcript.  `ukc` is talking to the federation-apiserver, and `kc` is talking to the normal cluster it's pointed at.

``` console
% tee >(cat) < ../data.kubernetes/cluster.yaml| ukc create -f -
apiVersion: federation/v1alpha1
kind: Cluster
metadata:
  name: federation-e2e-gce-us-central1-f
spec:
  secretRef:
    name: federation-e2e-gce-us-central1-f
  serverAddressByClientCIDRs:
  - clientCIDR: 0.0.0.0/0
    serverAddress: https://130.211.176.7
cluster "federation-e2e-gce-us-central1-f" created

% tee >(cat) < ../data.kubernetes/service.yaml| ukc create -f -
apiVersion: v1
kind: Service
metadata:
  name: e2e-foo
spec:
  selector:
    foo: bar
  ports:
    - name: portname1
      port: 80
      targetPort: svc1
    - name: portname2
      port: 80
      targetPort: svc2

service "e2e-foo" created

% ukc get services -oname
service/e2e-foo

% kc get services -oname
service/e2e-foo
service/kubernetes

% kc delete service e2e-foo
service "e2e-foo" deleted

% kc get services -oname -w
krun: Using _output/dockerized/bin
service/kubernetes
^C%
```

### PR DESCRIPTION
**Pull request by mfanjie**
fixes #27623 

[Analytics]()

### CODE REVIEW

**Comment by quinton-hoole** [comment]
ok to test

File: federation/pkg/federation-controller/service/servicecontroller.go [hunk 1]
```diff
@@ -47,7 +47,7 @@ import (
 )
 
 const (
-	serviceSyncPeriod = 10 * time.Minute
+	serviceSyncPeriod = 30 * time.Second
 	clusterSyncPeriod = 10 * time.Minute
```
Comment (nikhiljindal) [inline federation/pkg/federation-controller/service/servicecontroller.go]: cluster sync period should be updated as well?

**Comment by quinton-hoole** [comment]
@mfanjie This is going to cover up the fact that the watch is not triggering correctly, right?
I think that it would be better to continue to try to get to the root cause of the latter.  Agree?

**Comment by quinton-hoole** [comment]
I think that we should be watching each of the services in the underlying clusters, and be notified immediately (via a watch trigger) as soon as they are deleted (or change state).  Am I missing something?

**Comment by mfanjie** [comment]
@quinton-hoole, I agree with you that this is a cover up and we need to look for the real root cause. but per my understanding on the code flow, it gonna takes more times and risks at this critical. So I push this simplest PR to remove the limitation, I would suggest to fix it completely in v1.4.
@nikhiljindal cluster sync time does not impact on this problem, so I kept it as is.

**Comment by quinton-hoole** [comment]
@mfanjie could you in the mean time please confirm that we are at least
trying to watch the underlying services. If we're not, then we might have a
more serious (design) problem here.

Note that with say 100 clusters, each with (say) 1000 services in it, it's
clearly not feasible to poll each service every 30 seconds (or even 10
minutes), as there are 100,000 of them to get through.

**Comment by mfanjie** [comment]
@quinton-hoole currently we watches the underlying service status, if the service is deleted, we clean the  related info from cache. to fix this problem, after the cleanup, we need to get the federation service spec immediately, if it exists, we need to trigger service creation at once.
the side effect of this is federation service watcher might also see the changes and process another round of service creation. Potentially, there are multiple calls for same purpose. but finally, it will settle down.

**Comment by quinton-hoole** [comment]
OK, lets merge this as a safety measure, and in parallel try to debug why the watches are not firing correctly.  @mfanje can you take that on, and open an issue to track progress?

**Comment by quinton-hoole** [comment]
Sorry, I just noticed that you have pushed a completely different change in this commit.  The title of this issue needs to be updated accordingly.

I'll re-review your new changes now.

**Comment by quinton-hoole** [comment]
This new approach looks fine.  Thanks very much @mfanje!

**Pull request merged**